include/zephyr/net/net_if.h

```diff
@@ -2643,16 +2643,47 @@ void net_if_unregister_link_cb(struct net_if_link_cb *link);
 void net_if_call_link_cb(struct net_if *iface, struct net_linkaddr *lladdr,
 			 int status);
 
+/* used to ensure encoding of checksum support in net_if.h and
+ * ethernet.h is the same
+ */
+#define NET_IF_CHECKSUM_NONE_BIT			0
+#define NET_IF_CHECKSUM_IPV4_HEADER_BIT			BIT(0)
+#define NET_IF_CHECKSUM_IPV4_ICMP_BIT			BIT(1)
+#define NET_IF_CHECKSUM_IPV6_HEADER_BIT			BIT(10)
+#define NET_IF_CHECKSUM_IPV6_ICMP_BIT			BIT(11)
+#define NET_IF_CHECKSUM_TCP_BIT				BIT(21)
+#define NET_IF_CHECKSUM_UDP_BIT				BIT(22)
+
+/**
+ * @brief Type of checksum for which support in the interface will be queried.
+ */
+enum net_if_checksum_type {
+	NET_IF_CHECKSUM_IPV4_HEADER = NET_IF_CHECKSUM_IPV4_HEADER_BIT,
+	NET_IF_CHECKSUM_IPV4_TCP    = NET_IF_CHECKSUM_IPV4_HEADER_BIT |
```

Comment:
Is such granularity required?
TCP or UDP are IP agnostic (unlike ICMP). Hardware really have different support like that for l4 protocols, depending on l3?

Reply:
IPv4 and IPv6 have different header sizes. Also, for both protocols, the header size is variable and depends on the options selected (e.g., IPv6 has a number of optional headers that sit between the mandatory header and the transport header). The transport protocol used is also encoded in the IP header.
Thus, a NIC that does not understand the L3 protocol does not know which transport protocol is used, where the transport packet begins and where it ends and cannot insert a checksum for TCP or UDP.